mithril-common/src/chain_observer/pallas_observer.rs

```diff
@@ -175,6 +181,60 @@ impl PallasChainObserver {
         Ok(utxo)
     }
 
+    /// Fetches the current stake distribution using the provided `statequery` client.
+    async fn do_stake_snapshots_state_query(
+        &self,
+        statequery: &mut Client,
+    ) -> StdResult<StakeSnapshot> {
+        statequery
+            .acquire(None)
+            .await
+            .map_err(|err| anyhow!(err))
+            .with_context(|| "PallasChainObserver failed to acquire statequery")?;
+
+        let era = queries_v16::get_current_era(statequery)
+            .await
+            .map_err(|err| anyhow!(err))
+            .with_context(|| "PallasChainObserver failed to get current era")?;
+
+        let state_snapshot = queries_v16::get_stake_snapshots(statequery, era, BTreeSet::new())
+            .await
+            .map_err(|err| anyhow!(err))
+            .with_context(|| "PallasChainObserver failed to get stake snapshot")?;
+
+        Ok(state_snapshot)
+    }
+
+    /// Returns the stake pool hash from the given bytestring.
+    fn get_stake_pool_hash(&self, key: &Bytes) -> Result<String, ChainObserverError> {
+        let pool_hash = bech32::encode("pool", key.to_base32(), Variant::Bech32)
+            .map_err(|err| anyhow!(err))
+            .with_context(|| "PallasChainObserver failed to encode stake pool hash")?;
+
+        Ok(pool_hash)
+    }
+
+    /// Fetches the current stake distribution using the provided `statequery` client.
+    async fn get_stake_distribution(
+        &self,
+        client: &mut NodeClient,
+    ) -> Result<Option<StakeDistribution>, ChainObserverError> {
+        let statequery = client.statequery();
+
+        let stake_snapshot = self.do_stake_snapshots_state_query(statequery).await?;
+
+        let mut stake_distribution = StakeDistribution::new();
+
+        for (key, stakes) in stake_snapshot.snapshots.stake_snapshots.iter() {
+            if stakes.snapshot_mark_pool > 0 {
+                let pool_hash = self.get_stake_pool_hash(key)?;
+                stake_distribution.insert(pool_hash, stakes.snapshot_mark_pool);
+            }
+        }
```

Comment:
Can you refactor this bit like this:
```rust
let mut stake_distribution = StakeDistribution::new();

let have_stakes_in_two_epochs = |stakes: &Stakes| stakes.snapshot_mark_pool > 0;
for (key, stakes) in stake_snapshot
    .snapshots
    .stake_snapshots
    .iter()
    .filter(|(_, stakes)| have_stakes_in_two_epochs(stakes))
{
    let pool_hash = self.get_stake_pool_hash(key)?;
    stake_distribution.insert(pool_hash, stakes.snapshot_mark_pool);
}
``` 
This would makes more sense for someone that doesn't know much of the [Cardano inner mechanics here](https://cardano-course.gitbook.io/cardano-course/handbook/create-a-stake-pool/stake-snapshots).

Btw the tests below for are still green if we remove the `if stakes.snapshot_mark_pool`, you should add one or enhance the existing one for this case :).

Reply:
makes sense, thank you